src/lib/ledger_catchup/ledger_catchup.ml

```diff
@@ -120,118 +129,185 @@ module Make (Inputs : Inputs.S) :
+      ~preferred_peer ~maximum_download_size ~hashes_of_missing_transitions =
+    let random_peers = Network.random_peers network num_peers in
+    Deferred.Or_error.List.concat_map
+      (partition maximum_download_size hashes_of_missing_transitions)
+      ~how:`Sequential ~f:(fun hashes ->
+        Deferred.Or_error.find_map_ok (preferred_peer :: random_peers)
+          ~f:(fun peer ->
+            match%bind Network.get_transition_chain network peer hashes with
+            | None ->
+                Deferred.Or_error.errorf
+                  !"Peer %{sexp:Network_peer.Peer.t} did not have the \
+                    transitions we requested"
+                  peer
+            | Some transitions ->
+                if not @@ verify_against_hashes transitions hashes then (
+                  let error_msg =
+                    sprintf
+                      !"Peer %{sexp:Network_peer.Peer.t} returned a list that \
+                        is different from the one that is requested."
+                      peer
+                  in
+                  Trust_system.(
+                    record trust_system logger peer.host
+                      Actions.(Violated_protocol, Some (error_msg, [])))
+                  |> don't_wait_for ;
+                  Deferred.Or_error.error_string error_msg )
+                else
+                  Deferred.Or_error.return
+                  @@ List.map transitions ~f:(fun transition ->
+                         let transition_with_hash =
+                           With_hash.of_data transition
+                             ~hash_data:
+                               (Fn.compose Protocol_state.hash
+                                  External_transition.protocol_state)
+                         in
+                         Envelope.Incoming.wrap ~data:transition_with_hash
+                           ~sender:(Envelope.Sender.Remote peer.host) ) ) )
+
+  let verify_transitions_and_build_breadcrumbs ~logger ~trust_system ~verifier
+      ~frontier ~unprocessed_transition_cache ~transitions ~target_hash
+      ~subtrees =
     let open Deferred.Or_error.Let_syntax in
-    let%map result, initial_state_hash =
-      Deferred.Or_error.List.fold list ~init:([], None)
-        ~f:(fun (acc, initial_state_hash) elem ->
+    let%bind transitions_with_initial_validation, initial_hash =
+      fold_until (List.rev transitions) ~init:[]
+        ~f:(fun acc transition ->
           let open Deferred.Let_syntax in
-          if Option.is_some initial_state_hash then
-            Deferred.Or_error.return (acc, initial_state_hash)
+          match%bind
+            verify_transition ~logger ~trust_system ~verifier ~frontier
+              ~unprocessed_transition_cache transition
+          with
+          | Error e ->
+              List.map acc ~f:Cached.invalidate_with_failure |> ignore ;
+              Deferred.Or_error.fail e
+          | Ok (Either.First initial_hash) ->
```

Comment:
NIT: I forgot what was the data type that `First` and `Second` contained. Instead, can you use more explicit variants, like `In_froniter` and `Building_Path` or something.